File: frontend/src/api/income.ts
import http from './http'

export interface Account {
  id: number
  name: string
  sort_order: number
  is_active: boolean
}

export interface BalanceItem {
  account_id: number
  balance: number
}

export interface MonthlyBalanceRead {
  account_id: number
  account_name: string
  balance: number
  updated_at: string
}

export interface MonthlySnapshot {
  month: string
  total: number
  items: MonthlyBalanceRead[]
}

export interface TrendPoint {
  month: string
  value: number
}

export interface AccountTrend {
  account_id: number
  account_name: string
  data: TrendPoint[]
}

export const incomeApi = {
  // 账户列表（首次调用自动初始化默认账户）
  getAccounts() {
    return http.get<Account[]>('/income/accounts')
  },

  // 创建账户
  createAccount(name: string, sort_order = 0) {
    return http.post<Account>('/income/accounts', { name, sort_order })
  },

  // 更新账户
  updateAccount(id: number, data: { name?: string; sort_order?: number }) {
    return http.put<Account>(`/income/accounts/${id}`, data)
  },

  // 删除账户（软删除）
  deleteAccount(id: number) {
    return http.delete(`/income/accounts/${id}`)
  },

  // 所有月份快照列表（倒序）
  listMonths(limit = 24) {
    return http.get<MonthlySnapshot[]>('/income/balances', { params: { limit } })
  },

  // 单月详情
  getMonth(month: string) {
    return http.get<MonthlySnapshot>(`/income/balances/${month}`)
  },

  // 创建/更新某月余额（upsert）
  upsertBalances(month: string, balances: BalanceItem[]) {
    return http.post<MonthlySnapshot>('/income/balances', { month, balances })
  },

  // 删除某月全部记录
  deleteMonth(month: string) {
    return http.delete(`/income/balances/${month}`)
  },

  // 趋势数据（折线图）
  getTrend(months = 12) {
    return http.get<AccountTrend[]>('/income/stats/trend', { params: { months } })
  },

  // 导出 CSV
  exportCsv() {
    return http.get('/income/export/csv', { responseType: 'blob' })
  },

  // 导入 CSV
  importCsv(file: File) {
    const formData = new FormData()
    formData.append('file', file)
    return http.post<{ inserted: number; updated: number; skipped: number }>(
      '/income/import/csv',
      formData,
      { headers: { 'Content-Type': 'multipart/form-data' } },
    )
  },
}

File: backend/app/schemas/income.py
from datetime import datetime
from pydantic import BaseModel, Field


# ── 账户 CRUD ──────────────────────────────────────────────────────────────────

class AccountCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=32)
    sort_order: int = Field(0, ge=0)


class AccountUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=32)
    sort_order: int | None = Field(None, ge=0)


class AccountRead(BaseModel):
    id: int
    name: str
    sort_order: int
    is_active: bool

    model_config = {"from_attributes": True}


# ── 月度余额快照 ───────────────────────────────────────────────────────────────

class BalanceItem(BaseModel):
    """单个账户的余额"""
    account_id: int = Field(..., ge=1)
    balance: float = Field(..., ge=0)


class MonthlyBalanceUpsert(BaseModel):
    """创建/更新某月所有账户余额（批量）"""
    month: str = Field(..., pattern=r"^\d{4}-(0[1-9]|1[0-2])$", description="格式：YYYY-MM")
    balances: list[BalanceItem] = Field(..., min_length=1)


class MonthlyBalanceRead(BaseModel):
    account_id: int
    account_name: str
    balance: float
    updated_at: datetime

    model_config = {"from_attributes": True}


class MonthlySnapshotRead(BaseModel):
    """某月全部账户余额 + 总额"""
    month: str
    total: float
    items: list[MonthlyBalanceRead]


# ── 图表统计 ───────────────────────────────────────────────────────────────────

class TrendPoint(BaseModel):
    """折线图单个数据点"""
    month: str
    value: float


class AccountTrendRead(BaseModel):
    """单个账户的余额趋势"""
    account_id: int
    account_name: str
    data: list[TrendPoint]


class ChangePasswordRequest(BaseModel):
    old_password: str
    new_password: str
